sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala

```diff
@@ -551,3 +552,31 @@ case class ShowFunctions(
     pattern: Option[String]) extends Command {
   override def children: Seq[LogicalPlan] = child.toSeq
 }
+
+/**
+ * The logical plan of the ALTER TABLE ADD PARTITION command that works for v2 tables.
+ *
+ * The syntax of this command is:
+ * {{{
+ *     ALTER TABLE table ADD [IF NOT EXISTS]
+ *                 PARTITION spec1 [LOCATION 'loc1'][, PARTITION spec2 [LOCATION 'loc2'], ...];
+ * }}}
+ */
+case class AlterTableAddPartition(
+    table: SupportsPartitionManagement,
```

Comment:
You should make this `LogicalPlan` and create `AlterTableAddPartition(UnresolvedTableOrView, ...)` in `AstBuilder.scala`, and you can remove`*Statement` classes.

Then you can move the logics in `ResolveCatalog` to `DataSourceV2Strategy` using this class.

Reply:
The reason why I keep `AlterTableAddPartitionStatement` and `AlterTableAddPartition` is we still need analyze partitionIdents and make it qualified in `Analyzer`. It is different from RefreshTable, which only need to analyze UnresolvedTableOrView.

Reply:
This is a good point and does reveal a flaw in the current framework.

We can add some util classes to help us resolve partition spec
```
trait PartitionSpec

case class UnresolvedPartitionSpec(spec: TablePartitionSpec, location: Option[String]) extends PartitionSpec

case class ResolvedPartitionSpec(spec: InternalRow, location: Option[String]) extends PartitionSpec
```

then the command can be defined as
```
case class AlterTableAddPartition(table: LogicalPlan, partitions: Seq[PartitionSpec], ifExists: Boolean, ...)
```

And we add a rule to match `AlterTableAddPartition`, and resolve the `UnresolvedPartitionSpec` to `ResolvedPartitionSpec`.